# Stock Market Analysis
Stock Market Analysis means analyzing the current and historical trends in the stock market to make future buying and selling decisions. Stock market analysis is one of the best use cases of Data Science in finance. 

* Analysizing Current and Historical Trends
* Make Future buying and selling decisions 

# Data Collection
To analyze the stock market, I will collect the stock price data of Tata Consultancy Services to analyze the stock market interactively using the Python programming language. Let’s start by collecting the stock price data of Tata Consultancy Services. I will use the yfinance API of Yahoo Finance for collecting the stock price data. 

* Collecting Tata Consultancy Services Data

In [1]:
#intalling yfinance library
!pip install yfinance

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 63.2/63.2 kB 3.2 MB/s eta 0:00:00


In [2]:
import pandas as pd 
import yfinance as yf
import datetime
from datetime import date, timedelta
import plotly.graph_objects as go
import plotly.express as px

today = date.today()

d1 = today.strftime("%Y-%m-%d")
end_date = d1
d2 = date.today() - timedelta(days=365)
d2 = d2.strftime("%Y-%m-%d")
start_date = d2

data = yf.download('TCS.NS', 
                      start=start_date, 
                      end=end_date, 
                      progress=False)
data["Date"] = data.index
data = data[["Date", "Open", "High", "Low", 
             "Close", "Adj Close", "Volume"]]
data.reset_index(drop=True, inplace=True)
print(data.head())

        Date    Open         High          Low        Close    Adj Close  \
0 2022-06-27  3350.0  3362.899902  3306.500000  3314.699951  3208.111328   
1 2022-06-28  3307.0  3330.000000  3272.600098  3318.149902  3211.450439   
2 2022-06-29  3271.0  3333.000000  3271.000000  3290.500000  3184.689697   
3 2022-06-30  3272.0  3310.899902  3255.000000  3267.100098  3162.042236   
4 2022-07-01  3260.0  3333.000000  3252.050049  3315.100098  3208.498779   

    Volume  
0  2037686  
1  1400046  
2  2133994  
3  2719603  
4  1502879  


I convert stock data into DataFrame for my easy understanding.

In [3]:
df = pd.DataFrame(data)
df

Date         Open         High          Low        Close  \
0   2022-06-27  3350.000000  3362.899902  3306.500000  3314.699951   
1   2022-06-28  3307.000000  3330.000000  3272.600098  3318.149902   
2   2022-06-29  3271.000000  3333.000000  3271.000000  3290.500000   
3   2022-06-30  3272.000000  3310.899902  3255.000000  3267.100098   
4   2022-07-01  3260.000000  3333.000000  3252.050049  3315.100098   
..         ...          ...          ...          ...          ...   
242 2023-06-19  3190.000000  3214.899902  3175.000000  3210.399902   
243 2023-06-20  3214.000000  3232.500000  3201.050049  3227.699951   
244 2023-06-21  3250.000000  3261.399902  3223.949951  3258.199951   
245 2023-06-22  3260.000000  3267.649902  3230.100098  3238.500000   
246 2023-06-23  3237.300049  3249.399902  3209.149902  3216.350098   

       Adj Close   Volume  
0    3208.111328  2037686  
1    3211.450439  1400046  
2    3184.689697  2133994  
3    3162.042236  2719603  
4    3208.498779  1502879  
..           ...      ...  
242  3210.399902  1468186  
243  3227.699951  1528452  
244  3258.199951  1645987  
245  3238.500000  1529603  
246  3216.350098  1119505  

[247 rows x 7 columns]

In [4]:
round(df.describe(),2)

Open     High      Low    Close  Adj Close      Volume
count   247.00   247.00   247.00   247.00     247.00      247.00
mean   3256.35  3281.06  3227.83  3255.02    3192.99  1958799.43
std     122.05   120.90   123.49   122.79     134.52   970753.65
min    2959.85  3019.70  2926.10  2982.05    2893.78   260949.00
25%    3170.00  3190.60  3144.53  3172.90    3114.09  1372073.00
50%    3256.20  3278.15  3225.85  3256.70    3196.92  1719623.00
75%    3339.45  3368.85  3308.65  3334.93    3274.45  2216304.00
max    3549.00  3575.00  3527.00  3558.05    3531.78  6974600.00

# Find max close on Stock

In [5]:
df[df["Close" ]>3530]


Date         Open         High          Low        Close  \
156 2023-02-09  3515.000000  3547.600098  3505.300049  3540.850098   
157 2023-02-10  3520.000000  3542.350098  3508.149902  3537.550049   
161 2023-02-16  3527.199951  3575.000000  3527.000000  3558.050049   

       Adj Close   Volume  
156  3514.710693  1655427  
157  3511.435059  1187821  
161  3531.783691  1618931

**I can see maximum close in on date 2023-02-16 that is 3558.05**

In [6]:
figure = go.Figure(data=[go.Candlestick(x=data["Date"],
                                        open=data["Open"], high=data["High"],
                                        low=data["Low"], close=data["Close"])])
figure.update_layout(title = "Google Stock Price Analysis", xaxis_rangeslider_visible=False)
figure.show()

A bar plot is  a handy visualization to analyze the stock market, specifically in the long term. Here’s how to visualize the close prices of TCS stock using a bar plot:

In [7]:
figure = px.bar(data, x = "Date", y= "Close")
figure.show()

One of the valuable tools to analyze the stock market is a range slider. It helps to analyze the stock market between two specific points by interactively selecting the time period. 

In [8]:
figure = px.line(data, x='Date', y='Close', 
                 title='Stock Market Analysis with Rangeslider')
figure.update_xaxes(rangeslider_visible=True)
figure.show()

Another interactive feature I can add for stock market analysis is time period selectors. Time period selectors are like buttons that show you the graph of a specific time period. For example, a year, three months, six months, etc. Here is how you can add buttons for selecting the time period for stock market analysis:

In [9]:
figure = px.line(data, x='Date', y='Close', 
                 title='Stock Market Analysis with Time Period Selectors')

figure.update_xaxes(
    rangeselector=dict(
        buttons=list([
            dict(count=1, label="1m", step="month", stepmode="backward"),
            dict(count=6, label="6m", step="month", stepmode="backward"),
            dict(count=3, label="3m", step="month", stepmode="backward"),
            dict(count=1, label="1y", step="year", stepmode="backward"),
            dict(step="all")
        ])
    )
)
figure.show()

The weekend or holiday season always affects the stock market. So I want to remove all the records of the weekend trends from your stock market visualization :

In [10]:
figure = px.scatter(data, x='Date', y='Close', range_x=['2022-06-01', '2023-06-25'],
                 title="Stock Market Analysis by Hiding Weekend Gaps using scatter plot")
figure.update_xaxes(
    rangebreaks=[
        dict(bounds=["sat", "sun"])
    ]
)
figure.show()


In [11]:
figure = px.line(data, x='Date', y='Close', range_x=['2022-06-01', '2023-06-25'],
                 title="Stock Market Analysis by Hiding Weekend Gaps using line graph")
figure.update_xaxes(
    rangebreaks=[
        dict(bounds=["sat", "sun"])
    ]
)
figure.show()
